document/js-api/index.bs

```diff
@@ -325,74 +317,46 @@ A {{Module}} object represents a single WebAssembly module. Each {{Module}} obje
-            1. [=Append=] |table| to |tables|.
-        1. Otherwise:
-            1. Let |table| be the element in |tables| where |table|.\[[Table]] is |tableaddr|.
-        1. For each index |i| from 0 to the length of |table|.\[[Values]] &minus; 1, do
-            1. Let |funcaddr| be [=read_table=](|store|, |tableaddr|, |i|).
-            1. Assert: |funcaddr| is not [=error=] because this read is in bounds.
-            1. If |funcaddr| is not empty,
-                1. If |funcs| contains any |func| where |func|.\[[ExportedAddress]] is |funcaddr|,
-                    1. Let |func| be that function object.
-                1. Otherwise:
-                    1. Let |func| be a [=a new Exported Function=] created from |funcaddr|.
-                    1. [=Append=] |func| to |funcs|.
-                1. Set the |table|.\[[Values]][|i|] to |func|.
-
-                Note: The table and function objects created by the above steps are only observable for tables that are either imported or exported.
-
-    1. Let |startfuncaddr| be the start function of |instance|.
-    1. Let (|store|, |ret|) be [=invoke_func=](|store|, |startfuncaddr|, empty).
-    1. If |ret| is [=error=], throw an exception. This exception should be a WebAssembly {{RuntimeError}} exception, unless otherwise indicated by <a href="#errors">the WebAssembly error mapping</a>.
-    1. Assert: |ret| is empty.
+    1. Let (|store|, |instance|) be [=instantiate_module=](|store|, |module|, |imports|).
+    1. If |instance| is [=error=], throw an appropriate exception type:
+        * A {{LinkError}} exception for most cases which occur during linking.
+        * If the error came when running the start function, throw a {{RuntimeError}} for most errors which occur from WebAssembly, or the error object propagated from inner ECMAScript code.
+        * Another error type if appropriate, for example an out-of-memory exception, as documented in <a href="#errors">the WebAssembly error mapping</a>.
     1. Let |exportsObject| be ! [=ObjectCreate=](null).
-    1. For each pair (|nameBytes|, |externtype|) in [=module_exports=](|module|),
-        1. Let |name| be |nameBytes| [=UTF-8 decode without BOM or fail|decoded as UTF-8=].
-        1. Assert: |name| is not failure (|module| is [=valid=]).
+    1. For each pair (|name|, |externtype|) in [=module_exports=](|module|),
         1. Let |externval| be [=get_export=](|instance|, |name|).
+        1. Assert: |externval| is not [=error=].
         1. If |externtype| is of the form [=𝖿𝗎𝗇𝖼=] |functype|,
-            1. If |funcs| contains an entry |func| where 𝖿𝗎𝗇𝖼 |func|.\[[FunctionAddress]] is |externval|, let |value| be |func|
-            1. Otherwise,
-                1. Let |func| be the result of [=a new Exported Function=] from |externval|.
-                1. [=Append=] |func| to |funcs|.
-                1. Let |value| be |func|.
+            1. Then, |externval| is of the form [=external value|𝖿𝗎𝗇𝖼=] |funcaddr|.
```

Comment:
Turn this into an assertion?

Reply:
I was initially thinking an assertion, but is it OK for an assertion to introduce a binding? I'm not sure what the proper way to phrase this is for typical web spec conventions (cc @tobie).

Reply:
Hm, fair question. I think I have been using assertions for that in the core spec, but it's a fair point. Leaving it up to you.

Reply:
So an assertion doesn't do anything itself, it just asserts that at this point, that statement is true and would be a spec bug if it wasn't. See: https://infra.spec.whatwg.org/#assertions. In this particular context, I'm not too sure what "is of the form" actually means. (Maybe link it to a dfn?)

Reply:
The Wasm spec uses a sort of "algebraic data type" notation that I'm trying to "destructure" here. This differs from JS Records. I'm not sure what you would recommend. I could also imagine using "Let func funcaddr be externval" here.

Reply:
Oh, are you introducing |funcaddr| here for the first time? Then you should def to a `Let |funcaddr| be…` step.

Reply:
But, what would it be?

Reply:
So I'm not sure I understand precisely what's going on here (and I'm lacking context), which is precisely why I suggested using something like pr-preview to help with this.

I'm not sure I really understand how external values work well enough to comment.

Reply:
The cleanest solution probably is to split this into 2 lines: first the assert about the structure, then a destructuring let. I just checked through the core spec: destructuring let is used in a few places, but no cases exist where it might fail.